deepspeed/profiling/flops_profiler/README.md

```diff
@@ -35,216 +38,324 @@ Top 3 modules in flops at depth 2 are {'Conv2d': '421.91 MMACs', 'Linear': '11.1
+            2.36 M, 2.16% Params, 1.21 GMACs, 2.78% MACs, 9.26 ms, 2.37% time, 0.26 TFLOPS,
+            (dense): Linear(2.36 M, 2.16% Params, 1.21 GMACs, 2.78% MACs, 8.47 ms, 2.17% time, 0.29 TFLOPS, in_features=768, out_features=3072, bias=True)
+          )
+          (output): BertOutput(
+            2.36 M, 2.16% Params, 1.21 GMACs, 2.78% MACs, 5.81 ms, 1.49% time, 0.42 TFLOPS,
+            (dense): Linear(2.36 M, 2.16% Params, 1.21 GMACs, 2.78% MACs, 5.04 ms, 1.29% time, 0.48 TFLOPS, in_features=3072, out_features=768, bias=True)
+            (LayerNorm): LayerNorm(1.54 k, 0.00% Params, 0 MACs, 0.00% MACs, 342.61 us, 0.09% time, 0.0 TFLOPS, (768,), eps=1e-12, elementwise_affine=True)
+            (dropout): Dropout(0, 0.00% Params, 0 MACs, 0.00% MACs, 36.95 us, 0.01% time, 0.0 TFLOPS, p=0.1, inplace=False)
+          )
+        )
+        ...
+        (11): BertLayer(...)
+      )
+    )
+    (pooler): BertPooler(
+      590.59 k, 0.54% Params, 1.18 MMACs, 0.00% MACs, 264.41 us, 0.07% time, 0.0089 TFLOPS,
+      (dense): Linear(590.59 k, 0.54% Params, 1.18 MMACs, 0.00% MACs, 157.59 us, 0.04% time, 0.015 TFLOPS, in_features=768, out_features=768, bias=True)
+      (activation): Tanh(0, 0.00% Params, 0 MACs, 0.00% MACs, 42.44 us, 0.01% time, 0.0 TFLOPS, )
+    )
+  )
+  (dropout): Dropout(0, 0.00% Params, 0 MACs, 0.00% MACs, 19.79 us, 0.01% time, 0.0 TFLOPS, p=0.1, inplace=False)
+  (classifier): Linear(1.54 k, 0.00% Params, 3.07 KMACs, 0.00% MACs, 55.55 us, 0.01% time, 0.00011 TFLOPS, in_features=768, out_features=2, bias=True)
+)
+Top 3 modules in flops at depth 7 are {'Linear': '14.5 GMACs', 'Dropout': '0 MACs', 'LayerNorm': '0 MACs'}
+Top 3 modules in params at depth 7 are {'Linear': '28.35 M', 'LayerNorm': '18.43 k', 'Dropout': '0'}
+Top 3 modules in time at depth 7 are {'Linear': '149.14 ms', 'LayerNorm': '1.4 ms', 'Dropout': '641.82 us'}
+Number of multiply-adds:        43.49 GMACs
+Number of parameters:           109.48 M
+```
+
+### In Model Training Workflow
+
+To profile a model in a training workflow, use the `FlopsProfiler`class.
+The `FlopsProfiler`class provides the follwing methods:
+  * `start_profile()` - starts profiling
+  * `get_total_flops(as_string=False)` - returns the total number of MACs in the model
+  * `get_total_params(as_string=False)` - returns the total number of parameters in the model
+  * `print_model_profile()` - prints the model graph with the measured profile attached to each module
+  * `print_model_aggregated_profile(module_depth=-1, top_modules=3)` - prints the names of the top modules in terms of aggregated time, flops, and parameters at depth `module_depth`.
+  * `end_profile()` - ends profiling and cleans up. This should be invoked at the end of the profiling and after any printing method.
+
+#### Example Training Workflow
 
-# Output:
-# Number of multiply-adds:        21.74 GMACs
-# Number of parameters:           109.48 M
+Below is an example of this usage in a typical training workflow.
```

Comment:
In this mode, is the profiler capturing only the forward, or forward backward and step? Can we make this more explicit?

Reply:
The profiler only captures the forward. I clarify this through the README.